src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs

```diff
@@ -511,49 +512,120 @@ private void AutoCompleteCluster(object parameter)
+                    // Convert topology to actual cluster
+                    ClusterResultItem.Topology.Nodes.ToList().ForEach(node =>
+                    {
+                        // Retreive assembly name and node full name from type.id.
+                        var typeInfo = wsViewModel.NodeAutoCompleteSearchViewModel.GetInfoFromTypeId(node.Type.Id);
+                        wsViewModel.DynamoViewModel.Model.ExecuteCommand(new DynamoModel.CreateNodeCommand(Guid.NewGuid().ToString(), typeInfo.FullName, 0, 0, false, false));
+                        var nodeFromCluster = wsViewModel.Nodes.LastOrDefault();
+                        nodeFromCluster.IsTransient = true;
+                        clusterMapping.Add(node.Id, nodeFromCluster);
+                        // Add the node to the selection to prepare for autolayout later
+                        if (index == ClusterResultItem.EntryNodeIndex)
+                        {
+                            // This is the target node from cluster that should connect to the query node
+                            targetNodeFromCluster = nodeFromCluster;
+                        }
+                        index++;
+                    });
+
+                    ClusterResultItem.Topology.Connections.ToList().ForEach(connection =>
+                    {
+                        // Connect the nodes
+                        var sourceNode = clusterMapping[connection.StartNode.NodeId];
+                        var targetNode = clusterMapping[connection.EndNode.NodeId];
+                        // The port index is 1- based (currently a hack and not expected from service)
+                        var sourcePort = sourceNode.OutPorts.FirstOrDefault(p => p.PortModel.Index == connection.StartNode.PortIndex - 1);
+                        var targetPort = targetNode.InPorts.FirstOrDefault(p => p.PortModel.Index == connection.EndNode.PortIndex - 1);
+                        var commands = new List<DynamoModel.ModelBasedRecordableCommand>
+                        {
+                            new DynamoModel.MakeConnectionCommand(sourceNode.Id.ToString(), connection.StartNode.PortIndex - 1, PortType.Output, DynamoModel.MakeConnectionCommand.Mode.Begin),
+                            new DynamoModel.MakeConnectionCommand(targetNode.Id.ToString(), connection.EndNode.PortIndex - 1, PortType.Input, DynamoModel.MakeConnectionCommand.Mode.End),
+                        };
+                        commands.ForEach(c =>
+                        {
+                            try
+                            {
+                                wsViewModel.DynamoViewModel.Model.ExecuteCommand(c);
+                            }
+                            catch (Exception) { }
+                        });
+                    });
+
+                    // Connect the cluster to the original node and port
+                    var finalCommands = new List<DynamoModel.ModelBasedRecordableCommand>
```

Comment:
So I guess they need to be part of the same undo group. Currently if you undo it will go step by step for each node placement and connection.